### ISSUE TITLE
Prevent `let` and/or `const` from being used at too high of a scope

### ISSUE THREAD

**Issue by brettz9** [eslint/eslint#12992]
**Please describe what the rule should do:**

Should complain when a `const` or `let` variable is used at a higher depth than is necessary based on where it is used. Since `const` and `let` are block scoped, they can be meaningfully used at the lowest depth relevant. Doing so helps to clarify intent; when one sees a `const` or `let`, it would be helpful to know that this reliably signals that the variable is needed throughout that scope. Failing to use this forces attention on variables that are not of much global significance (i.e., they are only of interest to some nested block).

However, since `const` may be more frequently intended as file-wide , e.g., for an enum array, the rule ought to allow choosing on whether to allow on `let` and/or `const`.

And of course, in some situations, even a `let` variable may be meant as a tracker for use throughout a file, e.g., for counts, intended to be available in methods not yet added, but the rule could often simply be disabled in such cases, or, better yet, an additional option could be added, however, to allow variables at too high of a scope to nevertheless be permitted so long as the variable was used in at least two distinct descendant blocks. But generally speaking, I would expect that disabling the rule should not be so frequently necessary as the rule would check if it was needed across scopes, and only complain when used at an unnecessarily high scope.

`var` could perhaps be optionally enabled also, perhaps excepting complaining about the global scope (since it might be exported) and not going against the best practice (ala `vars-on-top`) by putting within a block but instead only moving it to a lower scope if there were a lower (function) scope to which it could move (at the top).

(Even `let` or `const` could be excepted for the global scope if used in a browser environment for the same reasons as [`no-implicit-globals`](https://eslint.org/docs/rules/no-implicit-globals) with `lexicalBindings` set to `true`, i.e., they can be global in a non-module environment.)

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
let i, l;
if (a) {
  for (i = 0; l = arr.length, i < l; i++) {

  }
}
```

...which could be auto-fixed to:

```js
if (a) {
  for (let i = 0; l = arr.length, i < l; i++) {

  }
}
```

```js
function hasFruit (fruit, a) {
  const possibleFruits = ['banana', 'orange', 'apple'];
  if (a) {
     return possibleFruits.includes(fruit);
  }
  doSomethingElse();
}
```

...which should instead be:
```js
function hasFruit (fruit, a) {
  if (a) {
     const possibleFruits = ['banana', 'orange', 'apple'];
     return possibleFruits.includes(fruit);
  }
  doSomethingElse();
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule is very fundamental, dealing with any kind of `const`/`let` assignments. It could help a large number of projects to avoid the need for mentally tracking variables that should not be of concern within a given scope. When many block-specific assignments are only present at lower scopes, developers can get a quicker higher-level picture of what is occurring at the higher level (e.g., other variables that actually are used to track state and what not throughout the whole scope). E.g.: instead of

```js
function aBigFunction () {
  let thisCanBe, too, crowded;
  const noGoodReasonToBeHere = [];
  if (cond) {
    //
  } else if (cond2) {
  }
}
```

..the following allows one to see very quickly that the main code is just concerned with two branches, and if one then determines one needs to only look inside the `cond` branch, then one is still not distracted with the other variables (and if one eventually comes to a low scope using these declarations, it is not too crowded, as the `const`/`let` keywords helps make clear at a glance that these are new variables being added whereas any without the keywords one would know must come from a higher scope).

```js
function aBigFunction () {
  if (cond) {
    //
  } elseif (cond2) {
    let thisIsNot, too, crowded;
    const goodReasonToBeHere = [];
  }
}
```
**Are you willing to submit a pull request to implement this rule?**

No, at least not at this time.

**Comment by mdjermanovic** [comment]
It seems complicated to reliably enforce this for `const` variables, which have to be initialized and are thus used in the same scope where they're declared:

```js
const a = foo();
bar();
if (cond) {
   // `const a = foo();` here might be different than above
   doSomething(a);
}
```

**Comment by brettz9** [comment]
Sure, good example, or for initialized let variables which use such complicated expressions. As with other rules, cases where auto-fixing could not be done would just be reported (or exempted).

**Comment by mdjermanovic** [comment]
> Sure, good example, or for initialized let variables which use such complicated expressions. As with other rules, cases where auto-fixing could not be done would just be reported, moved to the highest scope where no difference would occur, or exempted.

I think this isn't related just to auto-fixing. This could be a false positive, i.e. valid code where user can't change anything.

There are indeed some rules that would not auto-fix in some special cases (e.g., if it would remove a comment), but the reported error is still correct and can be fixed manually.

**Comment by brettz9** [comment]
Sure in such a case, it would need to be exempted from reporting not just fixing; that was one of the possibilities I mentioned. (I don't know that there are cases here where it was known to be problematic yet ambiguous as to how to fix; just mentioning it as one option.)

It would probably get too dicey to deal with the possibilities, but I just mean that there could be some cases which in theory could be auto-fixed, e.g., if `foo` was a pure function and wasn't passed variables used in that scope, etc., etc. I'm not suggesting one dive into such complexities, just that there could be some cases where it could still in theory be fixed.

**Comment by brettz9** [comment]
For starters, the rule might limit itself to `const`/`let` literal initializers (or empty `let` ones). That would probably be the bulk of cases anyways.

**Comment by kaicataldo** [comment]
I'm very hesitant to update existing or add new rules to core that have a high rate of false positives. This has been highly problematic in the past (see https://github.com/eslint/eslint/issues/11899), and I think it's generally indicative of an overreaching rule. ESLint has limitations (it doesn't know about types, runs on one file at a time, and can't know about side effects), and I don't think we should be trying to enforce anything that falls outside of these constraints (though we do have a number of existing rules that do).

**Comment by brettz9** [comment]
Sure, and maybe it was too ambitious to even bring up the possibility for analyzing anything more complex than literals or arrays and objects of literals. But as long as we're not moving out of the global scope in a non-module browser environment (since those `const`/`let` at top could be intended as exports in such environments), as long as we're not moving the fragile cases mentioned by @mdjermanovic (and the implementation does not fail to track variable usage or scopes), there should be no false positives.

While not a false positive, there could, however, be the already mentioned case of variables intended to remain at a higher scope despite their not being used within multiple lower scopes yet, but that can be minimized by the above-mentioned proposed config (that avoids de-hoisting when there are at least two separate scopes using the variable). But they wouldn't need to fear (and shouldn't) that the rule will break anything.

I think it may help to keep in mind that, while this can be a good practice in avoiding code clutter, this is--unlike #11899--also in a sense, more of a style rule. Some might not want to be forced to adhere to the style of de-hoisting, especially if they add variables before they are fully used, but this is not suggesting fixes that will cause breakage (projects that don't mind the occasional de-hoisting that they need to undo as they end up needing the de-hoisted variable can do so as part of normal development--as with many other rules such as `no-unused-vars`, `no-console`, etc. which are useful but which may cause some inevitable annoyances before a project is finalized).

**Comment by brettz9** [comment]
One other potential bonus we might add: avoiding some performance or temporary memory cost in building variables into memory...

**Comment by kaicataldo** [comment]
FWIW, this is a practice I generally try to follow in my own code, but I'm not sure ESLint is the best tool to enforce this given the constraints we have and our inability to always guess the user's intent. Do you have a solution for how we could solve the issue outlined [here](https://github.com/eslint/eslint/issues/12992#issuecomment-594536143) (i.e. what heuristic we could reliably use to ignore this case?).

**Comment by brettz9** [comment]
@kaicataldo : The heuristic would be to ignore such cases simply by detecting that the `init` of the `const a` declaration it is not a `Literal`, nor an `ArrayExpression` composed of `Literal`, nor an `ObjectExpression` whose `properties` are `Literal` (i.e., it does not belong to those expressions which are known to be safe to dehoist).

If such conditions are met, and if `a` is only placed at a lower scope where the intervening souces do not use the variable, it should be reliably safe to dehoist.

I.e., We would simply not attempt to analyze dynamic complex expressions which may or may not be safe to dehoist, and therefore wouldn't attempt to dehoist them.

So no fixers would be applied for cases like https://github.com/eslint/eslint/issues/12992#issuecomment-594536143 . Some users might want an option that *reported* such dynamic use, so that they could either disable the line, indicating it was intended or refactor where possible so that it could safely work at a lower level. This *option* would be like `require-atomic-updates` in that, despite its probable excessive loudness for regular users, it still does successfully warn you of cases you didn't want. However it would differ from `require-atomic-updates` in that such an option--assuming you wanted to even allow such an option--would be opt in. I'm not too big on having such an option (nor do I think one is needed for the rule to be useful), especially as it would be, unlike `require-atomic-updates`, not warning about any bugs.

But to be clear, my proposal does not seek to be a repeat of `require-atomic-updates` (except possibly with a non-default option).

**Comment by kaicataldo** [comment]
Thanks for explaining that. If this would only apply to hard coded constants, it sounds like this rule would only warn about the case in which the user is most likely to have intended to put the declaration at a higher scope than might be necessary, as a module-level constant.

**Comment by ljharb** [comment]
I very often put consts outside a function “unnecessarily” because i don’t want to re-evaluate the RHS every time the function runs. Would the rule warn on all of those?

**Comment by brettz9** [comment]
> [@kaicataldo]: Thanks for explaining that. If this would only apply to hard coded constants, it sounds like this rule would only warn about the case in which the user is most likely to have intended to put the declaration at a higher scope than might be necessary, as a module-level constant.

While I think the rule could still work if there is that config to only allow dehoisting when there are at least two uses of the variable in the document (ensuring it at least doesn't move too deep), I do realize from your comments that the constraints I mentioned would not be sufficient to cover the more frequent cases, yet I think it still ought to be possible to track some of the more frequent cases without adding too much complexity.

A frequent use of `const` I find is for the likes of this:

```js
const {var1, var2} = varParent;
```

In such a case, I think it would be sufficient to track that `varParent` is not altered in the intervening scopes, though since varParent could be altered by function calls within that scope, I guess it would either be necessary to avoid fixes if any function calls were intervening, or as feasible, to use any utilities you may be able to use for tracking variable use through transitive functions and such).

> [@ljharb]: I very often put consts outside a function “unnecessarily” because i don’t want to re-evaluate the RHS every time the function runs. Would the rule warn on all of those?

Good point. I would think that, excepting IIFEs (I would think any IIFE, not just `ExpressionStatement` ones should be ok), the dehoisting should not occur between function or loop scopes (except for `let` when there was no initializer to begin with). I think it would only safely or optimally work between conditionals, IIFEs, switches and the like.

Also, FWIW, I think there'd probably need to be config for cases like:

```js
const a = 5;
switch (b) {
    case 'sth':
        s = a + 7;
        break;
}

```

to add the braces like:

```js
switch (b) {
    case 'sth': {
        const a = 5;
        s = a + 7;
        break;
    }
}
```

Thank you both for helping analyze the potential pitfalls here, as the merits of the proposal are weighed. (I have some health issues giving me what is called "brain fog", and this can make it hard for me to work through more of the possibilities, though I know from experience that I have frequently run into this code I think is reasonable to wish to have auto-refactored (e.g., including, but not limited to, undoing the likes of `vars-on-top` and earlier JSLint code that enforced the same, for however good of a reason). And it is precisely because of my forgetfulness that I like `eslint` to be able to remind me of such things and help me declutter! :-)
